libmultipath/public_api.c

```diff
@@ -0,0 +1,794 @@
+}
+
+static void _fill_dmmp_mpath(struct dmmp_mpath *mpath, struct multipath *mpp,
+			     vector pathvec)
+{
+	char params[PARAMS_SIZE];
+	char status[PARAMS_SIZE];
+	int i = 0;
+	int j = 0;
+	struct pathgroup *pgp = NULL;
+	struct dmmp_path_group *mp_pg = NULL;
+	struct dmmp_path *mp_path = NULL;
+	struct path *pp = NULL;
+
+	if (pathvec == NULL)
+		goto error_out;
+
+	dm_get_map(mpp->alias, &mpp->size, params);
+	dm_get_status(mpp->alias, status);
+
+	disassemble_map(pathvec, params, mpp);
+	update_mpp_paths(mpp, pathvec);
+
+	_update_path_status(mpp);
+	mpp->bestpg = select_path_group(mpp);
+	disassemble_status(status, mpp);
+
+	if (VECTOR_SIZE(mpp->pg) == 0)
+		return;
+
+	mpath->mp_pgs = (struct dmmp_path_group **)
+		MALLOC(sizeof(struct dmmp_path_group *) * VECTOR_SIZE(mpp->pg));
+	mpath->mp_pg_count = VECTOR_SIZE(mpp->pg);
+
+	if (mpath->mp_pgs == NULL)
+		goto error_out;
+
+	vector_foreach_slot(mpp->pg, pgp, i) {
+		if (pgp == NULL)
+			goto error_out;
+		if (VECTOR_SIZE(pgp->paths) == 0)
+			goto error_out;
+		mp_pg = _dmmp_path_group_new();
+		if (mp_pg == NULL)
+			goto error_out;
+		mpath->mp_pgs[i] = mp_pg;
+		if (_fill_dmmp_path_group(mp_pg, pgp) != 0)
+			goto error_out;
+		if ((mp_pg->selector == NULL) && (mpp->selector != NULL)) {
+			/* Using struct multipath->selector is marked as
+			 * HACK in ./libmultipath/print.c.
+			 */
+			mp_pg->selector = STRDUP(mpp->selector);
+			if (mp_pg->selector == NULL)
+				goto error_out;
+		}
+
+		if (VECTOR_SIZE(pgp->paths) == 0)
+			continue;
+
+		mp_pg->mp_paths = (struct dmmp_path **)
+			MALLOC(sizeof(struct dmmp_path *) *
+			       VECTOR_SIZE(pgp->paths));
+		if (mp_pg->mp_paths == NULL)
+			goto error_out;
+		mp_pg->mp_path_count = VECTOR_SIZE(pgp->paths);
+
+		vector_foreach_slot(pgp->paths, pp, j) {
+			if (pp == NULL)
+				goto error_out;
+			mp_path = _dmmp_path_new();
+			if (mp_path == NULL)
+				goto error_out;
+			mp_pg->mp_paths[j] = mp_path;
+
+			mp_path->pg_id = mp_pg->id;
+			_fill_dmmp_path(mp_path, pp);
+		}
+
+	}
+	return;
+
+error_out:
+	if (mpath->mp_pgs != NULL)
+		_dmmp_path_group_list_free(mpath->mp_pgs, mpath->mp_pg_count);
+	mpath->mp_pgs = NULL;
+	mpath->mp_pg_count = 0;
+
+}
+
+static int _fill_dmmp_path_group(struct dmmp_path_group *mp_pg,
+				 struct pathgroup *pgp)
+{
+	struct path *pp = NULL;
+
+	pp = VECTOR_LAST_SLOT(pgp->paths);
+	mp_pg->id = pp->pgindex;
+	mp_pg->status = pgp->status;
+	if (pgp->priority < 0)
+		mp_pg->priority = 0;
+	else
+		mp_pg->priority = pgp->priority & UINT32_MAX;
+	if (pgp->selector != NULL) {
+		mp_pg->selector = STRDUP(pgp->selector);
+		if (mp_pg->selector == NULL)
+			return 1;
+	}
+	return 0;
+}
+
+static int _fill_dmmp_path(struct dmmp_path *mp_path, struct path *pp)
+{
+	strcpy(mp_path->name, pp->dev);
+	if (pp->state > 0)
+		mp_path->status = pp->state;
+	return 0;
+}
+
+static int _update_path_status(struct multipath *mpp)
+{
+	int i, j;
+	struct pathgroup * pgp;
+	struct path * pp;
+
+	if (!mpp->pg)
+		return 0;
```

Comment:
This method is copy from `multipath/main.c update_paths()`. I would like to keep it in
original form.